File: Samples/BasicApplication/BasicApplication/BasicApplication.Script/Northwind/Shared/PhoneEditor.cs
namespace BasicApplication.Northwind
{
    using jQueryApi;
    using Serenity;
    using Serenity.ComponentModel;
    using System;

    [Editor]
    public class PhoneEditor : StringEditor, IStringValue
    {
        public PhoneEditor(jQueryObject input)
            : base(input)
        {
            this.AddValidationRule(this.uniqueName, (e) =>
            {
                string value = this.Value.TrimToNull();
                if (value == null)
                    return null;

                return Validate(value, this.Multiple);
            });

            input.Bind("change", delegate(jQueryEvent e)
            {
                if (!e.HasOriginalEvent())
                    return;

                FormatValue();
            });

            input.Bind("blur", delegate(jQueryEvent e)
            {
                if (this.element.HasClass("valid"))
                {
                    FormatValue();
                }
            });
        }

        public void FormatValue()
        {
            this.element.Value(GetFormattedValue());
        }

        public string GetFormattedValue()
        {
            var value = this.element.GetValue();

            if (Multiple)
                return FormatMulti(value, FormatPhone);

            return FormatPhone(value);
        }

        [Option]
        public bool Multiple { get; set; }

        private static string Validate(string phone, bool isMultiple)
        {
            bool valid = isMultiple ? IsValidMulti(phone, IsValidPhone) : IsValidPhone(phone);

            if (valid)
                return null;

            return Q.Text(isMultiple ? "Validation.NorthwindPhoneMultiple" : "Validation.NorthwindPhone");
        }

        private static bool IsValidPhone(string phone)
        {
            if (phone.IsEmptyOrNull())
                return false;

            phone = phone.Replace(" ", "").Replace("-", "");

            if (phone.Length < 10)
                return false;

            if (phone.StartsWith("0"))
                phone = phone.Substring(1);

            if (phone.StartsWith("(") &&
                phone[4] == ')')
            {
                phone = phone.Substring(1, 3) + phone.Substring(5);
            }

            if (phone.Length != 10)
                return false;

            if (phone.StartsWith("0"))
                return false;

            for (var i = 0; i < phone.Length; i++)
            {
                var c = (int)phone[i];
                if (c < (int)'0' || c > (int)'9')
                    return false;
            }

            return true;
        }

        private static string FormatPhone(string phone)
        {
            if (!IsValidPhone(phone))
                return phone;

            phone = phone.Replace(" ", "").Replace("-", "").Replace("(", "").Replace(")", "");
            if (phone.StartsWith("0"))
                phone = phone.Substring(1);

            phone = "(" + phone.Substring(0, 3) + ") " + phone.Substring(3, 3) + "-" + phone.Substring(6, 2) + phone.Substring(8, 2);
            return phone;
        }

        private static string FormatMulti(string phone, Func<string, string> format)
        {
            var phones = phone.Replace(';', ',').Split(',');
            string result = "";
            foreach (var x in phones)
            {
                string s = x.TrimToNull();
                if (s == null)
                    continue;

                if (result.Length > 0)
                    result += ", ";

                result += format(s);
            }
            return result;
        }

        private static bool IsValidMulti(string phone, Func<string, bool> check)
        {
            if (phone.IsEmptyOrNull())
                return false;

            var phones = phone.Replace(';', ',').Split(',');
            bool anyValid = false;
            foreach (var x in phones)
            {
                string s = x.TrimToNull();
                if (s == null)
                    continue;

                if (!check(s))
                    return false;

                anyValid = true;
            }

            if (!anyValid)
                return false;

            return true;
        }

        public new string Value
        {
            get
            {
                return GetFormattedValue();
            }
            set
            {
                element.Value(value);
            }
        }
    }
}

File: common-features/src/demo.northwind/Modules/Shared/PhoneEditor.ts
import { Decorators, StringEditor, WidgetProps, localText, replaceAll } from "@serenity-is/corelib";

export interface PhoneEditorOptions {
    multiple?: boolean;
}

@Decorators.registerEditor('Serenity.Demo.Northwind.PhoneEditor')
export class PhoneEditor<P extends PhoneEditorOptions = PhoneEditorOptions> extends StringEditor<P> {

    constructor(props: WidgetProps<P>) {
        super(props);

        this.addValidationRule(this.uniqueName, e => {
            var value = this.get_value()?.trim();
            if (!value)
                return null;
            return PhoneEditor.validate(value, this.props?.multiple);
        });

        let input = this.domNode as HTMLInputElement;
        input.addEventListener('change', e => {
            this.formatValue();
        });

        input.addEventListener('blur', e => {
            if (this.domNode.classList.contains('valid')) {
                this.formatValue();
            }
        });
    }

    protected formatValue(): void {
        (this.domNode as HTMLInputElement).value = this.getFormattedValue();
    }

    protected getFormattedValue(): string {
        var value = (this.domNode as HTMLInputElement).value;
        if (this.props?.multiple) {
            return PhoneEditor.formatMulti(value, PhoneEditor.formatPhone);
        }
        return PhoneEditor.formatPhone(value);
    }

    get_value() {
        return this.getFormattedValue();
    }

    set_value(value: string) {
        (this.domNode as HTMLInputElement).value = value;
    }

    static validate(phone: string, isMultiple: boolean) {
        var valid = (isMultiple ? PhoneEditor.isValidMulti(phone, PhoneEditor.isValidPhone) : PhoneEditor.isValidPhone(phone));
        if (valid) {
            return null;
        }
        return localText((isMultiple ? 'Validation.NorthwindPhoneMultiple' : 'Validation.NorthwindPhone'));
    }

    static isValidPhone(phone: string) {
        if (!phone) {
            return false;
        }
        phone = replaceAll(replaceAll(phone, ' ', ''), '-', '');
        if (phone.length < 10) {
            return false;
        }

        if (phone.startsWith('0')) {
            phone = phone.substring(1);
        }

        if (phone.startsWith('(') && phone.charAt(4) === ')') {
            phone = phone.substring(1, 4) + phone.substring(5);
        }

        if (phone.length !== 10) {
            return false;
        }

        if (phone.startsWith('0')) {
            return false;
        }

        for (var i = 0; i < phone.length; i++) {
            var c = phone.charAt(i);
            if (c < '0' || c > '9') {
                return false;
            }
        }

        return true;
    }

    static formatPhone(phone) {
        if (!PhoneEditor.isValidPhone(phone)) {
            return phone;
        }
        phone = replaceAll(replaceAll(replaceAll(replaceAll(phone, ' ', ''), '-', ''), '(', ''), ')', '');
        if (phone.startsWith('0')) {
            phone = phone.substring(1);
        }
        phone = '(' + phone.substr(0, 3) + ') ' + phone.substr(3, 3) + '-' + phone.substr(6, 2) + phone.substr(8, 2);
        return phone;
    }

    static formatMulti(phone: string, format: (s: string) => string) {
        var phones = replaceAll(phone, String.fromCharCode(59), String.fromCharCode(44)).split(String.fromCharCode(44));
        var result = '';
        for (var x of phones) {
            var s = x?.trim();
            if (!s)
                continue;
            if (result.length > 0)
                result += ', ';
            result += format(s);
        }
        return result;
    }

    static isValidMulti(phone: string, check: (s: string) => boolean) {
        if (!phone)
            return false;
        var phones = replaceAll(phone, String.fromCharCode(59), String.fromCharCode(44)).split(String.fromCharCode(44));
        var anyValid = false;
        for (var $t1 = 0; $t1 < phones.length; $t1++) {
            var x = phones[$t1];
            var s = x?.trim();
            if (!s)
                continue;
            if (!check(s))
                return false;
            anyValid = true;
        }
        if (!anyValid)
            return false;
        return true;
    }
}
